homeassistant/components/homekit/type_sensors.py

```diff
@@ -274,16 +274,23 @@ def __init__(self, *args):
             else BINARY_SENSOR_SERVICE_MAP[DEVICE_CLASS_OCCUPANCY]
         )
 
+        self.format = service_char[2]
         service = self.add_preload_service(service_char[0])
-        self.char_detected = service.configure_char(service_char[1], value=0)
+        initial_value = False if self.format is bool else 0
+        self.char_detected = service.configure_char(
+            service_char[1], value=initial_value
+        )
         # Set the state so it is in sync on initial
         # GET to avoid an event storm after homekit startup
         self.update_state(state)
 
     def update_state(self, new_state):
         """Update accessory after state change."""
         state = new_state.state
-        detected = state in (STATE_ON, STATE_HOME)
+        if self.format == bool:
+            detected = state in (STATE_ON, STATE_HOME)
+        else:
+            detected = 1 if state in (STATE_ON, STATE_HOME) else 0
```

Comment:
Not saying it would improve readability, but this works too
```suggestion
        detected = self.format(state in (STATE_ON, STATE_HOME))
```

Reply:
WFM